File: src/ifcopenshell-python/ifcopenshell/util/placement.py
# IfcOpenShell - IFC toolkit and geometry engine
# Copyright (C) 2021 Dion Moult <dion@thinkmoult.com>
#
# This file is part of IfcOpenShell.
#
# IfcOpenShell is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# IfcOpenShell is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with IfcOpenShell.  If not, see <http://www.gnu.org/licenses/>.

import numpy as np
import numpy.typing as npt
import ifcopenshell
from typing import Literal, Optional
from collections.abc import Iterable

MatrixType = npt.NDArray[np.float64]
"""`npt.NDArray[np.float64]`"""


def a2p(o: Iterable[float], z: Iterable[float], x: Iterable[float]) -> MatrixType:
    """Converts a location, X, and Z axis vector to a 4x4 transformation matrix

    IFC uses a right-handed coordinate system, so it is not necessary to
    provide the Y axis.

    :param o: The origin (i.e. location) of the matrix
    :param z: The +Z vector / axis of the matrix
    :param x: The +X vector / axis of the matrix
    :return: A 4x4 numpy matrix
    """
    x = x / np.linalg.norm(x)
    z = z / np.linalg.norm(z)
    y = np.cross(z, x)
    y = y / np.linalg.norm(y)
    r = np.eye(4)
    r[:-1, :-1] = x, y, z
    r[-1, :-1] = o
    return r.T


def get_axis2placement(placement: ifcopenshell.entity_instance) -> MatrixType:
    """Parses an IfcAxis2Placement (2D or 3D) to a 4x4 transformation matrix

    Note that this function only parses a single placement axis. If you want to
    get the placement of an element instead, element placements often are made
    out of multiple placement axes or other alternative placement methods. You
    should use ``get_local_placement`` instead.

    :param placement: The IfcLocalPlacement enitity
    :return: A 4x4 numpy matrix
    """
    ifc_class = placement.is_a()
    if ifc_class in ("IfcAxis2Placement3D", "IfcAxis2PlacementLinear"):
        z = np.array(placement.Axis.DirectionRatios if placement.Axis else (0, 0, 1))
        x = np.array(placement.RefDirection.DirectionRatios if placement.RefDirection else (1, 0, 0))
        location = placement.Location
        if coordinates := getattr(location, "Coordinates", None):
            o = coordinates
        else:
            import ifcopenshell.geom

            settings = ifcopenshell.geom.settings()
            settings.set("convert-back-units", True)
            shape = ifcopenshell.geom.create_shape(settings, placement)
            return np.array(shape.matrix).reshape((4, 4), order="F")
    elif ifc_class == "IfcAxis2Placement2D":
        z = np.array((0, 0, 1))
        if placement.RefDirection:
            x = np.array(placement.RefDirection.DirectionRatios)
            x.resize(3)
        else:
            x = np.array((1, 0, 0))
        o = (*placement.Location.Coordinates, 0.0)

    elif ifc_class == "IfcAxis1Placement":
        axis = placement.Axis
        z = np.array(axis.DirectionRatios if axis else (0, 0, 1))
        x = np.array((1, 0, 0))
        o = placement.Location.Coordinates

    return a2p(o, z, x)


def get_local_placement(placement: Optional[ifcopenshell.entity_instance] = None) -> MatrixType:
    """Parse a local placement into a 4x4 transformation matrix

    This is typically used to find the location and rotation of an element. The
    transformation matrix takes the form of:

    .. code::

        [ [ x_x, y_x, z_x, x   ]
          [ x_y, y_y, z_y, y   ]
          [ x_z, y_z, z_z, z   ]
          [ 0.0, 0.0, 0.0, 1.0 ] ]

    Example:

    .. code:: python

        placement = file.by_type("IfcBeam")[0].ObjectPlacement
        matrix = ifcopenshell.util.placement.get_local_placement(placement)

    :param placement: The IfcLocalPlacement entity
    :return: A 4x4 numpy matrix
    """
    if placement is None:
        return np.eye(4)
    if (rel_to := placement.PlacementRelTo) is None:
        parent = np.eye(4)
    else:
        parent = get_local_placement(rel_to)
    return np.dot(parent, get_axis2placement(placement.RelativePlacement))


def get_cartesiantransformationoperator3d(inst: ifcopenshell.entity_instance) -> MatrixType:
    """Parses an IfcCartesianTransformationOperator into a 4x4 transformation matrix

    Note that in general you will not need to call this directly. See
    ``get_mappeditem_transformation`` instead.

    :param item: The IfcCartesianTransformationOperator entity
    :return: A 4x4 numpy transformation matrix
    """
    origin = np.array(inst.LocalOrigin.Coordinates)
    axis1 = np.array((1.0, 0.0, 0.0))
    axis2 = np.array((0.0, 1.0, 0.0))
    axis3 = np.array((0.0, 0.0, 1.0))

    if inst.Axis1:
        axis1[0:3] = inst.Axis1.DirectionRatios
    if inst.Axis2:
        axis2[0:3] = inst.Axis2.DirectionRatios
    if inst.Axis3:
        axis3[0:3] = inst.Axis3.DirectionRatios

    m4 = a2p(origin, axis3, axis1)
    # Negate axis2 (introduce mirroring) when supplied axis2
    # is opposite of constructed axis2, but remains orthogonal
    if m4.T[1][0:3].dot(axis2) < 0.0:
        m4.T[1] *= -1.0

    scale1 = scale2 = scale3 = 1.0

    if inst.Scale:
        scale1 = inst.Scale

    if inst.is_a("IfcCartesianTransformationOperator3DnonUniform"):
        scale2 = inst.Scale2 if inst.Scale2 is not None else scale1
        scale3 = inst.Scale3 if inst.Scale3 is not None else scale1
    else:
        scale2 = scale3 = scale1

    m4.T[0] *= scale1
    m4.T[1] *= scale2
    m4.T[2] *= scale3

    return m4


def get_mappeditem_transformation(item: ifcopenshell.entity_instance) -> MatrixType:
    """Parse an IfcMappedItem into a 4x4 transformation matrix

    Mapped items take a representation with an origin and transform them with a
    cartesian transformation operation. This function returns the final
    transformation matrix.

    :param item: The IfcMappedItem entity
    :return: A 4x4 numpy transformation matrix
    """
    m4 = get_axis2placement(item.MappingSource.MappingOrigin)
    # TODO 2d
    if item.MappingTarget.is_a("IfcCartesianTransformationOperator3D"):
        return get_cartesiantransformationoperator3d(item.MappingTarget) @ m4


def get_storey_elevation(storey: ifcopenshell.entity_instance) -> float:
    """Get the Z elevation in project units of a buildling storey

    Building storeys store elevation in two possible locations: the Z value of
    its placement, or as a fallback the ``Elevation`` attribute.

    :param storey: The IfcBuildingStorey entity
    :return: The elevation in project units
    """
    if storey.ObjectPlacement:
        matrix = get_local_placement(storey.ObjectPlacement)
        return matrix[2][3]
    return getattr(storey, "Elevation", 0.0) or 0.0


def rotation(angle: float, axis: Literal["X", "Y", "Z"], is_degrees=True) -> MatrixType:
    """Create a 4x4 numpy matrix representing an euler rotation

    :param angle: The angle of rotation
    :param axis: The axis to rotate around, either X, Y, or Z.
    :param is_degrees: Whether or not the angle is specified in degrees or
        radians. Defaults to true (i.e. degrees).
    :return: A 4x4 numpy rotation matrix
    """
    theta = np.radians(angle) if is_degrees else angle
    cos, sin = np.cos(theta), np.sin(theta)

    # fmt: off
    if axis == "X":
        return np.array([
            [1, 0, 0, 0],
            [0, cos, -sin, 0],
            [0, sin, cos, 0],
            [0, 0, 0, 1]
        ])
    elif axis == "Y":
        return np.array([
            [cos, 0, sin, 0],
            [0, 1, 0, 0],
            [-sin, 0, cos, 0],
            [0, 0, 0, 1]
        ])
    elif axis == "Z":
        return np.array([
            [cos, -sin, 0, 0],
            [sin, cos, 0, 0],
            [0, 0, 1, 0],
            [0, 0, 0, 1]
        ])
    # fmt: on

File: src/helpers/placement.cpp
/********************************************************************************
 *                                                                              *
 * This file is part of IfcOpenShell.                                           *
 *                                                                              *
 * IfcOpenShell is free software: you can redistribute it and/or modify         *
 * it under the terms of the Lesser GNU General Public License as published by  *
 * the Free Software Foundation, either version 3.0 of the License, or          *
 * (at your option) any later version.                                          *
 *                                                                              *
 * IfcOpenShell is distributed in the hope that it will be useful,              *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of               *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 *
 * Lesser GNU General Public License for more details.                          *
 *                                                                              *
 * You should have received a copy of the Lesser GNU General Public License     *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.         *
 *                                                                              *
 ********************************************************************************/

#include "placement.h"

#include "../ifcparse/exception.h"
#include "../ifcparse/instance_data.h"
#include "../ifcparse/schema.h"
#include "schema_dispatch.i"

#include <cstddef>
#include <type_traits>
#include <vector>

namespace {

Eigen::Vector3d safe_normalize(const Eigen::Vector3d& v,
                               const Eigen::Vector3d& fallback) {
    const double n = v.norm();
    return (n > 0.0) ? Eigen::Vector3d(v / n) : fallback;
}

template <typename Schema, typename = void>
struct has_axis2_placement_linear : std::false_type {};

template <typename Schema>
struct has_axis2_placement_linear<Schema, std::void_t<typename Schema::IfcAxis2PlacementLinear>> : std::true_type {};

[[noreturn]] void unsupported_schema(const std::string& name) {
    throw ifcopenshell::exception("No helper implementation was built for schema " + name);
}

template <typename Schema>
Eigen::Vector3d direction_or(const typename Schema::IfcDirection& direction,
                             const Eigen::Vector3d& fallback) {
    if (!direction) {
        return fallback;
    }
    const auto ratios = direction.DirectionRatios();
    if (ratios.size() < 2) {
        return fallback;
    }
    return Eigen::Vector3d(ratios[0], ratios[1], ratios.size() > 2 ? ratios[2] : 0.0);
}

template <typename Schema>
std::optional<Eigen::Vector3d> cartesian_point(const typename Schema::IfcPoint& point) {
    const auto cartesian = point.template as<typename Schema::IfcCartesianPoint>();
    if (!cartesian) {
        return std::nullopt;
    }
    const auto coordinates = cartesian.Coordinates();
    if (coordinates.size() < 2) {
        return std::nullopt;
    }
    return Eigen::Vector3d(coordinates[0], coordinates[1], coordinates.size() > 2 ? coordinates[2] : 0.0);
}

template <typename Schema, typename Placement>
Eigen::Matrix4d axis2_placement_3d_s(const Placement& placement) {
    const auto origin = cartesian_point<Schema>(placement.Location());
    if (!origin) {
        return Eigen::Matrix4d::Identity();
    }
    const auto z = direction_or<Schema>(placement.Axis(), Eigen::Vector3d::UnitZ());
    const auto x = direction_or<Schema>(placement.RefDirection(), Eigen::Vector3d::UnitX());
    return axes_to_placement(*origin, z, x);
}

template <typename Schema>
Eigen::Matrix4d get_axis2_placement_s(const express::base& placement) {
    if (auto axis3 = placement.template as<typename Schema::IfcAxis2Placement3D>()) {
        return axis2_placement_3d_s<Schema>(axis3);
    }
    if constexpr (has_axis2_placement_linear<Schema>::value) {
        if (auto linear = placement.template as<typename Schema::IfcAxis2PlacementLinear>()) {
            return axis2_placement_3d_s<Schema>(linear);
        }
    }
    if (auto axis2 = placement.template as<typename Schema::IfcAxis2Placement2D>()) {
        const auto origin = cartesian_point<Schema>(axis2.Location());
        if (!origin) {
            return Eigen::Matrix4d::Identity();
        }
        const auto x = direction_or<Schema>(axis2.RefDirection(), Eigen::Vector3d::UnitX());
        return axes_to_placement(*origin, Eigen::Vector3d::UnitZ(), x);
    }
    if (auto axis1 = placement.template as<typename Schema::IfcAxis1Placement>()) {
        const auto origin = cartesian_point<Schema>(axis1.Location());
        if (!origin) {
            return Eigen::Matrix4d::Identity();
        }
        const auto z = direction_or<Schema>(axis1.Axis(), Eigen::Vector3d::UnitZ());
        return axes_to_placement(*origin, z, Eigen::Vector3d::UnitX());
    }
    return Eigen::Matrix4d::Identity();
}

template <typename Schema>
Eigen::Matrix4d get_local_placement_s(const express::base& placement) {
    if (auto local = placement.template as<typename Schema::IfcLocalPlacement>()) {
        Eigen::Matrix4d parent = Eigen::Matrix4d::Identity();
        if (const auto relative_to = local.PlacementRelTo()) {
            parent = get_local_placement_s<Schema>(relative_to);
        }
        const auto relative = local.RelativePlacement();
        if (!relative) {
            return parent;
        }
        return parent * get_axis2_placement_s<Schema>(relative.concrete());
    }
    return get_axis2_placement_s<Schema>(placement);
}

// ifcopenshell.util.placement.get_storey_elevation: the Z of the storey's
// placement in project units, falling back to the Elevation attribute.
template <typename Schema>
double get_storey_elevation_s(const express::base& storey) {
    const auto typed = storey.template as<typename Schema::IfcBuildingStorey>();
    if (!typed) {
        return 0.0;
    }
    if (const auto placement = typed.ObjectPlacement()) {
        return get_local_placement_s<Schema>(placement)(2, 3);
    }
    // Fallback: the optional Elevation attribute (read by name).
    const ifcopenshell::entity* declaration = storey.declaration().as_entity();
    if (declaration != nullptr) {
        const std::ptrdiff_t index = declaration->attribute_index("Elevation");
        if (index >= 0) {
            const attribute_value value = storey.get_attribute_value(static_cast<std::size_t>(index));
            if (!value.isNull() && value.type() == ifcopenshell::Argument_DOUBLE) {
                return static_cast<double>(value);
            }
        }
    }
    return 0.0;
}

} // namespace

Eigen::Matrix4d axes_to_placement(const Eigen::Vector3d& origin,
                                  const Eigen::Vector3d& z,
                                  const Eigen::Vector3d& x) {
    const Eigen::Vector3d xn = safe_normalize(x, Eigen::Vector3d::UnitX());
    const Eigen::Vector3d zn = safe_normalize(z, Eigen::Vector3d::UnitZ());
    const Eigen::Vector3d yn = safe_normalize(zn.cross(xn), Eigen::Vector3d::UnitY());

    Eigen::Matrix4d m = Eigen::Matrix4d::Identity();
    m.block<3, 1>(0, 0) = xn;
    m.block<3, 1>(0, 1) = yn;
    m.block<3, 1>(0, 2) = zn;
    m.block<3, 1>(0, 3) = origin;
    return m;
}

Eigen::Matrix4d get_axis2_placement(const express::base& placement) {
    if (!placement) {
        return Eigen::Matrix4d::Identity();
    }
    const auto name = placement.declaration().schema()->name();
#define IFCOPENSHELL_DISPATCH(Schema, Identifier) \
    if (name == Identifier)                       \
        return get_axis2_placement_s<Schema>(placement);
    IFCOPENSHELL_HELPER_FOR_EACH_SCHEMA(IFCOPENSHELL_DISPATCH)
#undef IFCOPENSHELL_DISPATCH
    unsupported_schema(name);
}

Eigen::Matrix4d get_local_placement(const express::base& placement) {
    if (!placement) {
        return Eigen::Matrix4d::Identity();
    }
    const auto name = placement.declaration().schema()->name();
#define IFCOPENSHELL_DISPATCH(Schema, Identifier) \
    if (name == Identifier)                       \
        return get_local_placement_s<Schema>(placement);
    IFCOPENSHELL_HELPER_FOR_EACH_SCHEMA(IFCOPENSHELL_DISPATCH)
#undef IFCOPENSHELL_DISPATCH
    unsupported_schema(name);
}

double get_storey_elevation(const express::base& storey) {
    if (!storey) {
        return 0.0;
    }
    const auto name = storey.declaration().schema()->name();
#define IFCOPENSHELL_DISPATCH(Schema, Identifier) \
    if (name == Identifier)                       \
        return get_storey_elevation_s<Schema>(storey);
    IFCOPENSHELL_HELPER_FOR_EACH_SCHEMA(IFCOPENSHELL_DISPATCH)
#undef IFCOPENSHELL_DISPATCH
    unsupported_schema(name);
}
